### PR TITLE
Fix entity find path

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## Summary by CodeRabbit

- **New Features**
	- Introduced a more flexible, path-based entity search allowing users to navigate complex hierarchical structures with improved accuracy.
- **Bug Fixes**
	- Added validation to prevent unnecessary processing for empty path segments in entity searches.
- **Refactor**
	- Streamlined the underlying search logic to better handle edge cases, ensuring smoother and more reliable search results.
- **Tests**
	- Enhanced test coverage for the entity search functionality, including edge cases and clarifications for existing tests.

### CODE REVIEW

File: packages/core/src/Entity.ts [hunk 1]
```diff
@@ -22,15 +22,22 @@ export class Entity extends EngineObject {
...
+        // Search success if end path, or downward search
+        return isEndPath ? child : Entity._findChildByName(child, 0, paths, pathIndex + 1);
       }
     }
-    return null;
+
+    // Search failed if first path, or upward search
+    return pathIndex === 0
+      ? null
+      : Entity._findChildByName(entity.parent, entity.siblingIndex + 1, paths, pathIndex - 1);
```
Comment (luzhuang) [inline packages/core/src/Entity.ts:40]: siblingIndex over children.length, infinite loop

File: packages/core/src/Entity.ts [hunk 2]
```diff
@@ -22,15 +22,22 @@ export class Entity extends EngineObject {
   /**
    * @internal
    */
-  static _findChildByName(root: Entity, name: string): Entity {
-    const children = root._children;
-    for (let i = children.length - 1; i >= 0; i--) {
-      const child = children[i];
-      if (child.name === name) {
-        return child;
+  static _findChildByName(entity: Entity, childIndex: number, paths: string[], pathIndex: number): Entity {
```
Comment (cptbtptpbcptdtptp) [inline packages/core/src/Entity.ts:25]: It feels easier this way
```  
static _findChildByName(entity: Entity, paths: string[], index = 0): Entity {
    if (!entity || entity.name !== paths[index]) return null;
    if (index === paths.length - 1) return entity;
    const children = entity._children;
    for (let i = 0, n = children.length; i < n; i++) {
      const child = Entity._findChildByName(children[i], paths, index + 1);
      if (child) return child;
    }
    return null;
  }
```
